File: src/distiller_cm5_sdk/hardware/eink/test_display.py
#!/usr/bin/env python3
"""
Test script for e-ink display with configurable firmware.
"""

import sys
import os
from distiller_cm5_sdk.hardware.eink import (
    Display, DisplayMode, DisplayError, FirmwareType,
    set_default_firmware, get_default_firmware, initialize_display_config
)

def test_configuration():
    """Test configuration system."""
    print("=== Testing Configuration System ===")
    
    try:
        # Initialize config from environment/files
        print("1. Initializing configuration...")
        initialize_display_config()
        
        # Get current firmware
        current = get_default_firmware()
        print(f"Current firmware: {current}")
        
        # Test setting firmware
        print("2. Testing firmware setting...")
        if current == FirmwareType.EPD128x250:
            test_firmware = FirmwareType.EPD240x416
        else:
            test_firmware = FirmwareType.EPD128x250
        
        print(f"Setting firmware to: {test_firmware}")
        set_default_firmware(test_firmware)
        
        # Verify change
        updated = get_default_firmware()
        print(f"Updated firmware: {updated}")
        
        # Reset to original
        set_default_firmware(current)
        final = get_default_firmware()
        print(f"Reset to: {final}")
        
        return True
        
    except Exception as e:
        print(f"Configuration test failed: {e}")
        return False

def test_display_basic():
    """Test basic display operations."""
    print("\n=== Testing Basic Display Operations ===")
    
    try:
        # Create display without auto-initialization
        display = Display(auto_init=False)
        
        # Check if config is available
        if hasattr(display, '_config_available'):
            print(f"Config system available: {display._config_available}")
        
        # Initialize manually
        print("1. Initializing display...")
        display.initialize()
        
        # Get dimensions
        width, height = display.get_dimensions()
        print(f"Display dimensions: {width}x{height}")
        
        # Calculate correct data size
        array_size = (width * height) // 8
        print(f"Required data size: {array_size} bytes")
        
        # Test with correct data size
        print("2. Testing with correct data size...")
        
        # Create white image
        white_data = bytes([0xFF] * array_size)
        print(f"White image data: {len(white_data)} bytes")
        
        # Create black image
        black_data = bytes([0x00] * array_size)
        print(f"Black image data: {len(black_data)} bytes")
        
        # Display white image
        print("3. Displaying white image...")
        display.display_image(white_data, DisplayMode.FULL)
        print("✓ White image displayed")
        
        # Wait a bit
        import time
        time.sleep(2)
        
        # Display black image
        print("4. Displaying black image...")
        display.display_image(black_data, DisplayMode.FULL)
        print("✓ Black image displayed")
        
        time.sleep(2)
        
        # Clear display
        print("5. Clearing display...")
        display.clear()
        print("✓ Display cleared")
        
        # Sleep display
        print("6. Putting display to sleep...")
        display.sleep()
        print("✓ Display sleeping")
        
        # Cleanup
        display.close()
        print("✓ Display closed")
        
        return True
        
    except DisplayError as e:
        print(f"Display error: {e}")
        return False
    except Exception as e:
        print(f"Unexpected error: {e}")
        return False

def test_firmware_switch():
    """Test switching between firmware types."""
    print("\n=== Testing Firmware Switching ===")
    
    try:
        # Test with EPD128x250
        print("1. Testing with EPD128x250...")
        set_default_firmware(FirmwareType.EPD128x250)
        
        display = Display(auto_init=False)
        display.initialize()
        
        width, height = display.get_dimensions()
        print(f"EPD128x250 dimensions: {width}x{height}")
        
        expected_size = (width * height) // 8
        test_data = bytes([0xFF] * expected_size)
        
        # This should work
        display.display_image(test_data, DisplayMode.FULL)
        print(f"✓ EPD128x250 test passed ({expected_size} bytes)")
        
        display.close()
        
        # Test with EPD240x416 (if available)
        print("2. Testing with EPD240x416...")
        try:
            set_default_firmware(FirmwareType.EPD240x416)
            
            display = Display(auto_init=False)
            display.initialize()
            
            width, height = display.get_dimensions()
            print(f"EPD240x416 dimensions: {width}x{height}")
            
            expected_size = (width * height) // 8
            test_data = bytes([0xFF] * expected_size)
            
            # This should work too
            display.display_image(test_data, DisplayMode.FULL)
            print(f"✓ EPD240x416 test passed ({expected_size} bytes)")
            
            display.close()
            
        except Exception as e:
            print(f"EPD240x416 test failed (may not be available): {e}")
        
        return True
        
    except Exception as e:
        print(f"Firmware switch test failed: {e}")
        return False

def main():
    """Main test function."""
    print("E-ink Display Test Suite")
    print("=" * 50)
    
    success = True
    
    # Test configuration first
    if test_configuration():
        print("✓ Configuration test passed")
    else:
        print("✗ Configuration test failed")
        success = False
    
    # Test basic display operations
    if test_display_basic():
        print("✓ Basic display test passed")
    else:
        print("✗ Basic display test failed")
        success = False
    
    # Test firmware switching
    if test_firmware_switch():
        print("✓ Firmware switch test passed")
    else:
        print("✗ Firmware switch test failed")
        success = False
    
    print("\n" + "=" * 50)
    if success:
        print("✅ All tests passed!")
        return 0
    else:
        print("❌ Some tests failed!")
        return 1

if __name__ == "__main__":
    sys.exit(main())

File: src/distiller_cm5_sdk/hardware/eink/lib/src/bin/test_display.rs
#!/usr/bin/env rust-script

use distiller_display_sdk_shared::{
    config, FirmwareType, DisplayMode,
    display_init, display_image_raw, display_clear, display_sleep, display_cleanup,
    create_white_image, create_black_image, get_dimensions
};
use std::env;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize logging
    env_logger::init();
    
    println!("=== E-ink Display Test ===");
    
    // Initialize configuration system
    println!("1. Initializing configuration...");
    if let Err(e) = config::initialize_config() {
        println!("Warning: Config initialization failed: {}", e);
        println!("Using default configuration");
    }
    
    // Get current firmware
    match config::get_default_firmware() {
        Ok(firmware) => println!("Current firmware: {}", firmware),
        Err(e) => println!("Error getting firmware: {}", e),
    }
    
    // Check for environment variable override
    if let Ok(firmware_env) = env::var("DISTILLER_EINK_FIRMWARE") {
        println!("Setting firmware from environment: {}", firmware_env);
        if let Err(e) = config::set_default_firmware_from_str(&firmware_env) {
            println!("Error setting firmware: {}", e);
        } else {
            match config::get_default_firmware() {
                Ok(firmware) => println!("Updated firmware: {}", firmware),
                Err(e) => println!("Error getting updated firmware: {}", e),
            }
        }
    }
    
    // Get display dimensions
    println!("\n2. Getting display dimensions...");
    let dimensions = get_dimensions();
    println!("Display dimensions: {}x{} pixels", dimensions.0, dimensions.1);
    
    let array_size = (dimensions.0 * dimensions.1 / 8) as usize;
    println!("Required data size: {} bytes", array_size);
    
    // Test image creation
    println!("\n3. Testing image creation...");
    let white_image = create_white_image();
    let black_image = create_black_image();
    
    println!("White image size: {} bytes", white_image.len());
    println!("Black image size: {} bytes", black_image.len());
    
    if white_image.len() != array_size {
        println!("ERROR: Image size mismatch!");
        println!("Expected: {} bytes, got: {} bytes", array_size, white_image.len());
        return Err("Image size mismatch".into());
    }
    
    // Test display initialization
    println!("\n4. Testing display initialization...");
    match display_init() {
        Ok(()) => {
            println!("✓ Display initialized successfully");
            
            // Test display operations
            println!("\n5. Testing display operations...");
            
            // Display white image
            println!("Displaying white image...");
            if let Err(e) = display_image_raw(&white_image, DisplayMode::Full) {
                println!("Error displaying white image: {}", e);
            } else {
                println!("✓ White image displayed");
            }
            
            std::thread::sleep(std::time::Duration::from_secs(2));
            
            // Display black image
            println!("Displaying black image...");
            if let Err(e) = display_image_raw(&black_image, DisplayMode::Full) {
                println!("Error displaying black image: {}", e);
            } else {
                println!("✓ Black image displayed");
            }
            
            std::thread::sleep(std::time::Duration::from_secs(2));
            
            // Clear display
            println!("Clearing display...");
            if let Err(e) = display_clear() {
                println!("Error clearing display: {}", e);
            } else {
                println!("✓ Display cleared");
            }
            
            // Sleep display
            println!("Putting display to sleep...");
            if let Err(e) = display_sleep() {
                println!("Error putting display to sleep: {}", e);
            } else {
                println!("✓ Display sleeping");
            }
            
            // Cleanup
            println!("Cleaning up...");
            if let Err(e) = display_cleanup() {
                println!("Error during cleanup: {}", e);
            } else {
                println!("✓ Cleanup completed");
            }
            
        }
        Err(e) => {
            println!("✗ Display initialization failed: {}", e);
            println!("This could be due to:");
            println!("  - Hardware not connected");
            println!("  - Insufficient permissions (try with sudo)");
            println!("  - SPI/GPIO devices not available");
            println!("  - Wrong firmware configuration");
            return Err(e.into());
        }
    }
    
    println!("\n=== Test completed successfully ===");
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_config_system() {
        // Test firmware type parsing
        assert!(config::FirmwareType::from_str("EPD128x250").is_ok());
        assert!(config::FirmwareType::from_str("EPD240x416").is_ok());
        assert!(config::FirmwareType::from_str("invalid").is_err());
    }
    
    #[test]
    fn test_image_creation() {
        // Test with default firmware
        let white = create_white_image();
        let black = create_black_image();
        
        assert_eq!(white.len(), black.len());
        assert!(white.len() > 0);
        
        // All bytes in white image should be 0xFF
        assert!(white.iter().all(|&b| b == 0xFF));
        
        // All bytes in black image should be 0x00
        assert!(black.iter().all(|&b| b == 0x00));
    }
    
    #[test]
    fn test_dimensions() {
        let dims = get_dimensions();
        assert!(dims.0 > 0);
        assert!(dims.1 > 0);
        
        // Should be one of the supported resolutions
        let supported = [(128, 250), (240, 416)];
        assert!(supported.contains(&dims));
    }
}